app/sample/gpu_direct/tx_st20_pipeline_gpu_direct.c

```diff
@@ -0,0 +1,279 @@
+    close(fd);
+    return -EIO;
+  }
+  m = mmap(NULL, i.st_size, PROT_READ, MAP_SHARED, fd, 0);
+  if (MAP_FAILED == m) {
+    err("%s, mmap %s failed\n", __func__, file);
+    close(fd);
+    return -EIO;
+  }
+  frame_cnt = i.st_size / s->frame_size;
+  fbs_size = i.st_size;
+  info("%s, tx_url %s frame_cnt %d\n", __func__, file, frame_cnt);
+
+init_fb:
+
+  s->source_begin = mtl_hp_zmalloc(s->st, fbs_size, MTL_PORT_P);
+  if (!s->source_begin) {
+    err("%s, source malloc on hugepage failed\n", __func__);
+    if (m) munmap(m, i.st_size);
+    if (fd >= 0) close(fd);
+    return -EIO;
+  }
+  s->frame_cursor = s->source_begin;
+  if (m) mtl_memcpy(s->source_begin, m, fbs_size);
+  s->source_end = s->source_begin + fbs_size;
+
+  if (m) munmap(m, i.st_size);
+  if (fd >= 0) close(fd);
+
+  return 0;
+}
+
+static int tx_st20p_frame_done(void* priv, struct st_frame* frame) {
+  struct tx_st20p_sample_ctx* s = priv;
+  MTL_MAY_UNUSED(frame);
+
+  s->fb_send_done++;
+  dbg("%s(%d), done %d\n", __func__, s->idx, s->fb_send_done);
+  return 0;
+}
+
+static void tx_st20p_build_frame(struct tx_st20p_sample_ctx* s, struct st_frame* frame) {
+  uint8_t* src = s->frame_cursor;
+
+  mtl_memcpy(frame->addr[0], src, s->frame_size);
+}
+
+static void* tx_st20p_frame_thread(void* arg) {
+  struct tx_st20p_sample_ctx* s = arg;
+  st20p_tx_handle handle = s->handle;
+  struct st_frame* frame;
+
+  info("%s(%d), start\n", __func__, s->idx);
+  while (!s->stop) {
+    frame = st20p_tx_get_frame(handle);
+    if (!frame) { /* no frame */
+      warn("%s(%d), get frame time out\n", __func__, s->idx);
+      continue;
+    }
+
+    if (s->source_begin) tx_st20p_build_frame(s, frame);
+    if (s->has_user_meta) {
+      s->meta.idx = s->fb_send;
+      frame->user_meta = &s->meta;
+      frame->user_meta_size = sizeof(s->meta);
+    }
+    st20p_tx_put_frame(handle, frame);
+
+    /* point to next frame */
+    s->frame_cursor += s->frame_size;
+    if (s->frame_cursor + s->frame_size > s->source_end) {
+      s->frame_cursor = s->source_begin;
+    }
+    s->fb_send++;
+    dbg("%s(%d), fb_send %d\n", __func__, s->idx, s->fb_send);
+  }
+  info("%s(%d), stop\n", __func__, s->idx);
+
+  return NULL;
+}
+
+int main(int argc, char** argv) {
+  struct st_sample_context ctx = {0};
+  int ret;
+
+  /* init sample(st) dev */
+  ret = tx_sample_parse_args(&ctx, argc, argv);
+  if (ret < 0) return ret;
+
+  /* enable auto start/stop */
+  ctx.param.flags |= MTL_FLAG_DEV_AUTO_START_STOP;
+  ctx.st = mtl_init(&ctx.param);
+  if (!ctx.st) {
+    err("%s: mtl_init failed\n", __func__);
+    return -EIO;
+  }
+
+  uint32_t session_num = ctx.sessions;
+  struct tx_st20p_sample_ctx* app[session_num];
+
+  /* create and register tx session */
+  for (int i = 0; i < session_num; i++) {
+    app[i] = calloc(1, sizeof(struct tx_st20p_sample_ctx));
+    if (!app[i]) {
+      err("%s(%d), app context memory allocation failed\n", __func__, i);
+      ret = -ENOMEM;
+      goto error;
```

Comment:
Exiting from the loop with `goto`, really?

Reply:
We will fix it as part of overall effort to fix "goto" handling in all sample application pipelines.

Reply:
Consider the next technique
```
bool err = false;
for (...) {
   ...
   if (something) {
      err = true;
      break;
   }
   ...
}
if (err)
   goto error;
```